src/regdesk/printer_manager.js

```diff
@@ -0,0 +1,171 @@
+import { LabelEpl, LP2844 } from 'WebZLP/src/LP2844'
+import { PrinterDropdown } from './printer_dropdown.js';
+
+export class PrinterManager {
+
+  #adultDropdown;
+  #minorDropdown;
+  #dropdowns;
+
+  #printerType = LP2844;
+
+  /**
+   * Initializes a new instance of the PrinterManager class.
+   *
+   * @param {Navigator} nav - Browser Navigator to hook up events to.
+   * @param {PrinterDropdown} adultDropdown - The dropdown object managing the adult printer.
+   * @param {PrinterDropdown} minorDropdown - The dropdown object managing the Minor printer.
+   */
+  constructor(nav, adultDropdown, minorDropdown) {
+    this.#adultDropdown = adultDropdown;
+    this.#adultDropdown.addEventListener(PrinterDropdown.ConnectEventName, async (e) => this.handleRequestConnect(e));
+    this.#adultDropdown.addEventListener(PrinterDropdown.DisconnectEventName, async (e) => this.handleRequestDisconnect(e));
+
+    this.#minorDropdown = minorDropdown;
+    this.#minorDropdown.addEventListener(PrinterDropdown.ConnectEventName, async (e) => this.handleRequestConnect(e));
+    this.#minorDropdown.addEventListener(PrinterDropdown.DisconnectEventName, async (e) => this.handleRequestDisconnect(e));
+
+    // Convenience array
+    this.#dropdowns = [this.#adultDropdown, this.#minorDropdown];
+
+    nav.usb.addEventListener('connect', async (e) => this.handleConnectPrinter(e));
+    nav.usb.addEventListener('disconnect', async (e) => this.handleDisconnectPrinter(e));
+  }
+
+  /**
+   * Assign a printer to a dropdown
+   * @param {LP2844} printer
+   * @param {PrinterDropdown} dropdown
+   */
+  assignPrinter(printer, dropdown) {
+    // Find the first dropdown in order, hope it's the right one!
+    // TODO: Make this smarter and pull from a cache or something for assignment
```

Comment:
I agree with you. But I don't think we need a cache or anything for assignment! Rather, the problem exists because we have two different methods trying to resolve this situation. Specifically refreshPrinters separates the dropdown from the printer, and assignPrinter tries to join them back together. But I think if refreshPrinters is doing the separation, it should be the one to do the join.

```
assignPrinter(printer, dropdown) {
  dropdown.setPrinter(printer);
}

async refreshPrinters() {
  const dropdownByPrinterSerial = await Promise.all(this.#dropdowns.map(async (dropdown) => {
    const serial = dropdown.printer.serial;
    await dropdown.removePrinter();
    return {serial: d.dropdownName};
  });

  return navigator.usb.getDevices().then(async (devices) => {
    return Promise.all(devices.map(async device => {
      const printer = await this.handleConnectPrinter({ device });
      const dropdown = this.#dropdowns.find(dropdown => dropdown.dropdownName === dropdownByPrinterSerial[printer.serial]);
      this.assignPrinter(printer, dropdown);
    });
  });
}
```

(This also fixes a number of promise issues, see my comments below).

Reply:
Woops, that first promise.all returns an array, not an object map. But I'm sure you figured that out. :)